FUSETOOLS2-2154: Add Set Exchange property test

unsortedhashsets opened the pull request:
Hi, added the test according to the issue https://issues.redhat.com/browse/FUSETOOLS2-2154
   
Camel Debugger tests
    ✔ Toggle breakpoint on log line (17) (40658ms)
    ✔ Update Body value with Camel debugger (15664ms)
    ✔ Update Header value with Camel debugger (13338ms)
    ✔ Update Exchange property value with Camel debugger (3673ms)
    ✔ Click on Continue button, and check updated message (13461ms)
    ✔ Untoggle breakpoint on log line (17) (448ms)
 
 Extra:
- Changed template file (property added)
- Changed common variables to reflect file changes
- Separate reload test to three test for each changed element (body, header, property)

**Question/Possible Issue:**

I faced a lack of logs in `camel.impl.debugger.BacklogDebugger`

Example for Message/Body value:
2023-07-25 12:44:12.944  INFO 47051 --- [on(4)-127.0.0.1] camel.impl.debugger.BacklogDebugger : Breakpoint at node log1 is updating message body on exchangeId: EC2A49CFFCBC3AA-0000000000000000 with new body: TEST

But for Exchange properties for example newly added or CamelTimerPeriod, CamelTimerCounter
No logs in terminal in case of value changed

Variables (Changed `from`, `CamelTimerPeriod`, `Body` :
![Screenshot from 2023-07-25 12-47-19](https://github.com/camel-tooling/camel-dap-client-vscode/assets/46084942/3c93bc7e-e242-4f5d-9cbb-66461d18857d)

Logs:
![Screenshot from 2023-07-25 12-48-36](https://github.com/camel-tooling/camel-dap-client-vscode/assets/46084942/cbb8f082-d696-4a54-81ee-2f5d5242f939)

---

Extra  to avoid the error in root: 
   - Deleted redundant (always true) expect with getText() method
   - Added clean util to clean terminal
   - Increased log timeout and terminal check from 1s/0.5s to 2 seconds

djelinek commented on src/ui-test/tests/debugger.test.ts:
```diff
@@ -24,10 +24,12 @@ import {
 } from '../utils';
 
-describe('Camel Debugger tests', function () {
+describe.only('Camel Debugger tests', function () {
```
TODO: remove `only`

unsortedhashsets replied:
:facepalm:

djelinek commented:
regarding your question @unsortedhashsets . I am afraid I have a lack of camel debugger knowledge to explain this behaviour. I would say feel free to open separate issue for it and we can wait for @apupier and discuss in another thread what is happening there.

unsortedhashsets commented:
Some issues appeared on Ubuntu machines:

  1) "after all" hook in "{root}"

  28 passing (10m)
  1 failing

  1) "after all" hook in "{root}":
     Error: Timeout of 60000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
      at listOnTimeout (node:internal/timers:559:17)
      at processTimers (node:internal/timers:502:7)

Need to investigate more

unsortedhashsets commented:
> Some issues appeared on Ubuntu machines:
> 
> 1. "after all" hook in "{root}"
> 
> 28 passing (10m) 1 failing
> 
> 1. "after all" hook in "{root}":
>    Error: Timeout of 60000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
>    at listOnTimeout (node:internal/timers:559:17)
>    at processTimers (node:internal/timers:502:7)
> 
> Need to investigate more

Temporarily solved with:

- Deleted redundant (always true) expect with getText() method
- Added clean util to clean terminal
- Increased log timeout and terminal check from 1s/0.5s to 2 seconds

The true solution needs some optimization on the extester side

djelinek commented on src/ui-test/tests/02_camel.settings.test.ts:
```diff
@@ -63,13 +62,11 @@ describe('Camel User Settings', function () {
             await executeCommand(CAMEL_RUN_ACTION_LABEL);
+            
+            // Static wait due initial camel instalations through the jbang
+            await driver.sleep(70000);
```
🙁

djelinek replied:
is this really needed? that is not nice at all 🙁

unsortedhashsets replied:
Not needed, but it always will be (the first camel jbang run installs a lot of things and just wait for the response for 30 seconds) and dynamic wait with getText will run +-35 times with the negative result, the idea was to limit the call of getText() for now

To see it locally you need to clean jbang cache:

1) `jbang cache clear -o --all`
2) run some dap command

unsortedhashsets replied:
deleted for now, but if we will face that "limit" issue again (in case no fix on extester side) it will be the first place to optimize :+1:

djelinek commented:
1 code smell reported by Sonarcloud - https://sonarcloud.io/project/issues?id=camel-tooling_camel-dap-client-vscode&pullRequest=453&resolved=false&types=CODE_SMELL

djelinek commented on src/ui-test/tests/debugger.test.ts:
```diff
@@ -108,24 +131,84 @@ describe('Camel Debugger tests', function () {
 
-    it('Untoogle breakpoint on log line (14)', async function () {
+    it('Untoggle breakpoint on log line (17)', async function () {
+        if(!breakpointToggled){this.test?.skip();}
```
why not to use condition for `it` on line 208?

```
 (breakpointToggled ? it : it.skip)
```

unsortedhashsets replied:
Tested

![Screenshot from 2023-08-01 16-26-47](https://github.com/camel-tooling/camel-dap-client-vscode/assets/46084942/25d1bb06-ec4a-47bc-bb16-19c484cea438)

djelinek replied:
I meant more like `(breakpointToggled ? it : it.skip)('Untoggle breakpoint on log line (17)', async function () {`

unsortedhashsets replied:
:open_mouth: wow

unsortedhashsets replied:
now sonar counts all file conditions :rofl: 
`Refactor this function to reduce its Cognitive Complexity from 55 to the 15 allowed.`

djelinek replied:
sonar was better when did not worked 😆

djelinek replied:
we can consider and change manually in conarcloud as false positive for this condition for `it`

unsortedhashsets replied:
Seems that way does not work with changed variables...only with const (as we have for macOS)

djelinek commented on src/ui-test/tests/debugger.test.ts:
```diff
@@ -54,15 +63,20 @@ describe('Camel Debugger tests', function () {
     });
 
     it('Update Body value with Camel debugger', async function () {
+        if (skip){ 
```
I don't like it much, but I see the point that other tests are connected and all will fail anyway in case previous ones not pass..

maybe at least visually could also ternary operator looks a bit more elegant WDYT?

`skip ? this.test?skip() : skip = true;`

unsortedhashsets replied:
It was recognized by sonar as "CODE_SMELL" :smile: previously

djelinek replied:
ok, sonarcloud is the most smart guy in the room then

unsortedhashsets replied:
![Screenshot from 2023-08-01 16-26-35](https://github.com/camel-tooling/camel-dap-client-vscode/assets/46084942/e3f53158-7624-42e1-80e1-81c810300b3b)

unsortedhashsets replied:
compromise between the three of us =)

```
if (skip){ this.test?.skip(); }
skip = true;
```

djelinek replied:
I don't have strong opinion here.. choose more readable and clean notation

unsortedhashsets replied:
I will use the same as for breakpoint :open_mouth:

djelinek commented on src/ui-test/utils.ts:
```diff
@@ -88,15 +92,21 @@ export async function selectContextMenuItem(command: string, menu: ContextMenu):
  */
-export async function waitUntilTerminalHasText(driver: WebDriver, textArray: string[], interval = 500): Promise<void> {
+export async function waitUntilTerminalHasText(
+            driver: WebDriver,
```
interesting code formatting

unsortedhashsets commented on src/ui-test/tests/reload.test.ts:
```diff
@@ -70,18 +75,25 @@ describe('Jbang commands with automatic reload', function () {
     });
 
-    afterEach(async function () {
-        await killTerminal();
```
in reload test now it is in after section
